File: MiniDOMFinanziaria/Finanziaria/StatisticheLavorazione/CQSPD.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Finanziaria
Imports minidom.Sistema
Imports minidom.Anagrafica

Partial Public Class Finanziaria

    <Serializable>
    Public Class CPraticaInfoLavorazione
        Implements minidom.XML.IDMDXMLSerializable

        Public m_IDPratica As Integer
        <NonSerialized> Public m_Pratica As CPraticaCQSPD
        Public IDPersona As Integer
        <NonSerialized> Public Persona As CPersonaFisica
        Public NomePersona As String
        Public Rata As Decimal
        Public Durata As Integer
        Public NettoRicavo As Decimal
        Public NettoAllaMano As Decimal
        Public TAN As Single
        Public TAEG As Single

        Public Sub New()
            Me.m_IDPratica = 0
            Me.m_Pratica = Nothing
            Me.IDPersona = 0
            Me.Persona = Nothing
            Me.NomePersona = ""
            Me.Rata = 0
            Me.Durata = 0
            Me.NettoAllaMano = 0
            Me.NettoRicavo = 0
            Me.TAEG = 0
            Me.TAN = 0
        End Sub

        Public Sub New(ByVal p As CPraticaCQSPD)
            If (p Is Nothing) Then Throw New ArgumentNullException("p")
            Me.m_IDPratica = GetID(p)
            Me.m_Pratica = p
            Me.Persona = p.Cliente
            Me.IDPersona = GetID(Me.Persona)
            Me.NomePersona = Me.Persona.Nominativo
            Me.Rata = p.OffertaCorrente.Rata
            Me.Durata = p.OffertaCorrente.Durata
            Me.NettoRicavo = p.OffertaCorrente.NettoRicavo
            'Me.m_NettoAllaMano = p.OffertaCorrente.NettoRicavo - p.altr
            Me.TAN = p.OffertaCorrente.TAN
            Me.TAEG = p.OffertaCorrente.TAEG
        End Sub

        Public Property IDPratica As Integer
            Get
                Return GetID(Me.m_Pratica, Me.m_IDPratica)
            End Get
            Set(value As Integer)
                If (Me.IDPratica = value) Then Exit Property
                Me.m_IDPratica = value
                Me.m_Pratica = Nothing
            End Set
        End Property

        Public Property Pratica As CPraticaCQSPD
            Get
                If (Me.m_Pratica Is Nothing) Then Me.m_Pratica = Finanziaria.Pratiche.GetItemById(Me.m_IDPratica)
                Return Me.m_Pratica
            End Get
            Set(value As CPraticaCQSPD)
                Me.m_Pratica = value
                Me.m_IDPratica = GetID(value)
            End Set
        End Property



        Public Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "IDPratica" : Me.m_IDPratica = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "IDPersona" : Me.IDPersona = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NomePersona" : Me.NomePersona = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Rata" : Me.Rata = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "Durata" : Me.Durata = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NettoRicavo" : Me.NettoRicavo = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "NettoAllaMano" : Me.NettoAllaMano = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "TAN" : Me.TAN = XML.Utils.Serializer.DeserializeDouble(fieldValue)
                Case "TAEG" : Me.TAEG = XML.Utils.Serializer.DeserializeDouble(fieldValue)
            End Select
        End Sub

        Public Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("IDPratica", Me.IDPratica)
            writer.WriteAttribute("IDPersona", Me.IDPersona)
            writer.WriteAttribute("NomePersona", Me.NomePersona)
            writer.WriteAttribute("Rata", Me.Rata)
            writer.WriteAttribute("Durata", Me.Durata)
            writer.WriteAttribute("NettoRicavo", Me.NettoRicavo)
            writer.WriteAttribute("NettoAllaMano", Me.NettoAllaMano)
            writer.WriteAttribute("TAN", Me.TAN)
            writer.WriteAttribute("TAEG", Me.TAEG)
        End Sub
    End Class


    <Serializable>
    Public Class CQSPDSTXANNSTATITEM
        Implements IComparable, XML.IDMDXMLSerializable

        Public Anno As Integer
        Public Conteggio() As Integer
        Public ML() As Decimal

        Public Sub New()
            Me.Anno = 0
            ReDim Me.Conteggio(12)
            ReDim Me.ML(12)
        End Sub

        Public Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
            Dim tmp As CQSPDSTXANNSTATITEM = obj
            Dim ret As Integer = Me.Anno - tmp.Anno
            Return ret
        End Function

        Protected Overridable Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Anno" : Me.Anno = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Conteggio" : Me.Conteggio = XML.Utils.Serializer.ToArray(Of Integer)(fieldValue)
                Case "ML" : Me.ML = XML.Utils.Serializer.ToArray(Of Integer)(fieldValue)
            End Select
        End Sub

        Protected Overridable Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Anno", Me.Anno)
            writer.WriteTag("Conteggio", Me.Conteggio)
            writer.WriteTag("ML", Me.ML)
        End Sub
    End Class


#Region "Statistiche per prodotto"

    Public Enum InfoStatoEnum As Integer
        CONTATTO = 0
        LIQUIDATO = 1
        ANNULLATO = 2
        ALTRO = 3
    End Enum

    Public Class InfoStato
        Implements XML.IDMDXMLSerializable

        Public Descrizione As String
        Public stato As InfoStatoEnum
        Public Conteggio As Integer
        Public Montante As Decimal

        Public Sub New(ByVal descrizione As String, ByVal stato As InfoStatoEnum, ByVal conteggio As Integer, ByVal montante As Decimal)
            DMDObject.IncreaseCounter(Me)
            Me.Descrizione = descrizione
            Me.stato = stato
            Me.Conteggio = conteggio
            Me.Montante = montante
        End Sub

        Protected Overridable Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Descrizione" : Me.Descrizione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Stato" : Me.stato = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Conteggio" : Me.Conteggio = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Montante" : Me.Montante = XML.Utils.Serializer.DeserializeDouble(fieldValue)
            End Select
        End Sub

        Protected Overridable Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Descrizione", Me.Descrizione)
            writer.WriteAttribute("Stato", Me.stato)
            writer.WriteAttribute("Conteggio", Me.Conteggio)
            writer.WriteAttribute("Montante", Me.Montante)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

    Public Class CompareByAnnullato
        Implements IComparer, IComparer(Of CRigaStatisticaPerStato)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Public Function Compare(x As CRigaStatisticaPerStato, y As CRigaStatisticaPerStato) As Integer Implements IComparer(Of CRigaStatisticaPerStato).Compare
            Dim v1 As Decimal = x.Item(InfoStatoEnum.ANNULLATO).Montante
            Dim v2 As Decimal = y.Item(InfoStatoEnum.ANNULLATO).Montante
            If (v1 < v2) Then Return 1
            If (v1 > v2) Then Return -1
            Return 0
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function
    End Class

    Public Class CompareByContatto
        Implements IComparer, IComparer(Of CRigaStatisticaPerStato)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Public Function Compare(x As CRigaStatisticaPerStato, y As CRigaStatisticaPerStato) As Integer Implements IComparer(Of CRigaStatisticaPerStato).Compare
            Dim v1 As Decimal = x.Item(InfoStatoEnum.CONTATTO).Montante
            Dim v2 As Decimal = y.Item(InfoStatoEnum.CONTATTO).Montante
            If (v1 < v2) Then Return 1
            If (v1 > v2) Then Return -1
            Return 0
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function
    End Class

    Public Class CompareByLiquidato
        Implements IComparer, IComparer(Of CRigaStatisticaPerStato)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Public Function Compare(x As CRigaStatisticaPerStato, y As CRigaStatisticaPerStato) As Integer Implements IComparer(Of CRigaStatisticaPerStato).Compare
            Dim v1 As Decimal = x.Item(InfoStatoEnum.LIQUIDATO).Montante
            Dim v2 As Decimal = y.Item(InfoStatoEnum.LIQUIDATO).Montante
            If (v1 < v2) Then Return 1
            If (v1 > v2) Then Return -1
            Return 0
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function
    End Class

    Public Class CompareByAltriStati
        Implements IComparer, IComparer(Of CRigaStatisticaPerStato)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Public Function Compare(x As CRigaStatisticaPerStato, y As CRigaStatisticaPerStato) As Integer Implements IComparer(Of CRigaStatisticaPerStato).Compare
            Dim v1 As Decimal = x.Item(InfoStatoEnum.ALTRO).Montante
            Dim v2 As Decimal = y.Item(InfoStatoEnum.ALTRO).Montante
            If (v1 < v2) Then Return 1
            If (v1 > v2) Then Return -1
            Return 0
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function
    End Class

    Public Class CompareByKey
        Implements IComparer, IComparer(Of CRigaStatisticaPerStato)

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub

        Public Function Compare(x As CRigaStatisticaPerStato, y As CRigaStatisticaPerStato) As Integer Implements IComparer(Of CRigaStatisticaPerStato).Compare
            Return Strings.Compare(x.Descrizione, y.Descrizione, CompareMethod.Text)
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function
    End Class

    <Serializable>
    Public Class CRigaStatisticaPerStato
        Implements IComparable, XML.IDMDXMLSerializable

        Public Descrizione As String = ""
        Public m_Items() As InfoStato
        Public Tag As String
        Public Tag1 As Integer

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
            ReDim Me.m_Items(3)
            Me.m_Items(0) = New InfoStato("Contatto", InfoStatoEnum.CONTATTO, 0, 0)
            Me.m_Items(1) = New InfoStato("Liquidato", InfoStatoEnum.LIQUIDATO, 0, 0)
            Me.m_Items(2) = New InfoStato("Annullato", InfoStatoEnum.ANNULLATO, 0, 0)
            Me.m_Items(3) = New InfoStato("Altri Stati", InfoStatoEnum.ALTRO, 0, 0)
        End Sub

        Public ReadOnly Property Item(ByVal stato As InfoStatoEnum) As InfoStato
            Get
                Select Case stato
                    Case InfoStatoEnum.ALTRO : Return Me.m_Items(3)
                    Case InfoStatoEnum.ANNULLATO : Return Me.m_Items(2)
                    Case InfoStatoEnum.LIQUIDATO : Return Me.m_Items(1)
                    Case Else : Return Me.m_Items(0)
                End Select
            End Get
        End Property

        Public Function CompareTo(ByVal b As CRigaStatisticaPerStato) As Integer
            Return Strings.Compare(Me.Descrizione, b.Descrizione, CompareMethod.Text)
        End Function

        Private Function CompareTo1(obj As Object) As Integer Implements IComparable.CompareTo
            Return Me.CompareTo(obj)
        End Function

        Protected Overridable Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
            Select Case fieldName
                Case "Descrizione" : Me.Descrizione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Tag" : Me.Tag = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Tag1" : Me.Tag1 = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Items" : Me.m_Items = XML.Utils.Serializer.ToArray(Of InfoStato)(fieldValue)
            End Select
        End Sub

        Protected Overridable Sub XMLSerialize(writer As XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
            writer.WriteAttribute("Descrizione", Me.Descrizione)
            writer.WriteAttribute("Tag", Me.Tag)
            writer.WriteAttribute("Tag1", Me.Tag1)
            writer.WriteTag("Items", Me.m_Items)
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

    <Serializable>
    Public Class CRigaStatisticaPerStatoCollection
        Inherits CKeyCollection(Of CRigaStatisticaPerStato)

        Public Sub New()
        End Sub

        Public Shadows Function Add(ByVal descrizione As String) As CRigaStatisticaPerStato
            Dim item As New CRigaStatisticaPerStato
            item.Descrizione = descrizione
            MyBase.Add("" & descrizione, item)
            'MyBase.Sort()
            Return item
        End Function

    End Class

#End Region



End Class

File: MiniDOMFinanziaria/Finanziaria/StatisticheLavorazione/CQSPD.cs
using System;
using System.Collections;
using System.Collections.Generic;
using DMD;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class Finanziaria
    {
        [Serializable]
        public class CPraticaInfoLavorazione
            : IDMDXMLSerializable
        {
            public int m_IDPratica;
            [NonSerialized] public CPraticaCQSPD m_Pratica;
            public int IDPersona;
            [NonSerialized] public Anagrafica.CPersonaFisica Persona;
            public string NomePersona;
            public decimal Rata;
            public int Durata;
            public decimal NettoRicavo;
            public decimal NettoAllaMano;
            public float TAN;
            public float TAEG;

            public CPraticaInfoLavorazione()
            {
                m_IDPratica = 0;
                m_Pratica = null;
                IDPersona = 0;
                Persona = null;
                NomePersona = "";
                Rata = 0m;
                Durata = 0;
                NettoAllaMano = 0m;
                NettoRicavo = 0m;
                TAEG = 0f;
                TAN = 0f;
            }

            public CPraticaInfoLavorazione(CPraticaCQSPD p)
            {
                if (p is null)
                    throw new ArgumentNullException("p");
                m_IDPratica = DBUtils.GetID(p);
                m_Pratica = p;
                Persona = p.Cliente;
                IDPersona = DBUtils.GetID(Persona);
                NomePersona = Persona.Nominativo;
                Rata = p.OffertaCorrente.Rata;
                Durata = p.OffertaCorrente.Durata;
                NettoRicavo = p.OffertaCorrente.NettoRicavo;
                // Me.m_NettoAllaMano = p.OffertaCorrente.NettoRicavo - p.altr
                TAN = (float)p.OffertaCorrente.TAN;
                TAEG = (float)p.OffertaCorrente.TAEG;
            }

            public int IDPratica
            {
                get
                {
                    return DBUtils.GetID(m_Pratica, m_IDPratica);
                }

                set
                {
                    if (IDPratica == value)
                        return;
                    m_IDPratica = value;
                    m_Pratica = null;
                }
            }

            public CPraticaCQSPD Pratica
            {
                get
                {
                    if (m_Pratica is null)
                        m_Pratica = Pratiche.GetItemById(m_IDPratica);
                    return m_Pratica;
                }

                set
                {
                    m_Pratica = value;
                    m_IDPratica = DBUtils.GetID(value);
                }
            }

            public void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "IDPratica":
                        {
                            m_IDPratica = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "IDPersona":
                        {
                            IDPersona = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "NomePersona":
                        {
                            NomePersona = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Rata":
                        {
                            Rata = (decimal)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "Durata":
                        {
                            Durata = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "NettoRicavo":
                        {
                            NettoRicavo = (decimal)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "NettoAllaMano":
                        {
                            NettoAllaMano = (decimal)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "TAN":
                        {
                            TAN = (float)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }

                    case "TAEG":
                        {
                            TAEG = (float)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }
                }
            }

            public void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("IDPratica", IDPratica);
                writer.WriteAttribute("IDPersona", IDPersona);
                writer.WriteAttribute("NomePersona", NomePersona);
                writer.WriteAttribute("Rata", Rata);
                writer.WriteAttribute("Durata", Durata);
                writer.WriteAttribute("NettoRicavo", NettoRicavo);
                writer.WriteAttribute("NettoAllaMano", NettoAllaMano);
                writer.WriteAttribute("TAN", TAN);
                writer.WriteAttribute("TAEG", TAEG);
            }
        }

        [Serializable]
        public class CQSPDSTXANNSTATITEM 
            : IComparable, IDMDXMLSerializable
        {
            public int Anno;
            public int[] Conteggio;
            public decimal[] ML;

            public CQSPDSTXANNSTATITEM()
            {
                Anno = 0;
                Conteggio = new int[13];
                ML = new decimal[13];
            }

            public int CompareTo(object obj)
            {
                CQSPDSTXANNSTATITEM tmp = (CQSPDSTXANNSTATITEM)obj;
                int ret = Anno - tmp.Anno;
                return ret;
            }

            protected virtual void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Anno":
                        {
                            Anno = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Conteggio":
                        {
                            Conteggio = (int[])DMD.XML.Utils.Serializer.ToArray<int>(fieldValue);
                            break;
                        }

                    case "ML":
                        {
                            ML = (decimal[])DMD.XML.Utils.Serializer.ToArray<int>(fieldValue);
                            break;
                        }
                }
            }

            protected virtual void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Anno", Anno);
                writer.WriteTag("Conteggio", Conteggio);
                writer.WriteTag("ML", ML);
            }

            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue) { this.SetFieldInternal(fieldName, fieldValue);  }
            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer) { this.XMLSerialize(writer);  }
        }


        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        public enum InfoStatoEnum : int
        {
            CONTATTO = 0,
            LIQUIDATO = 1,
            ANNULLATO = 2,
            ALTRO = 3
        }

        public class InfoStato 
            : IDMDXMLSerializable
        {
            public string Descrizione;
            public InfoStatoEnum stato;
            public int Conteggio;
            public decimal Montante;

            public InfoStato(string descrizione, InfoStatoEnum stato, int conteggio, decimal montante)
            {
                DMDObject.IncreaseCounter(this);
                Descrizione = descrizione;
                this.stato = stato;
                Conteggio = conteggio;
                Montante = montante;
            }

            protected virtual void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Descrizione":
                        {
                            Descrizione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Stato":
                        {
                            stato = (InfoStatoEnum)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Conteggio":
                        {
                            Conteggio = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Montante":
                        {
                            Montante = (decimal)DMD.XML.Utils.Serializer.DeserializeDouble(DMD.Strings.CStr(fieldValue));
                            break;
                        }
                }
            }

            protected virtual void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Descrizione", Descrizione);
                writer.WriteAttribute("Stato", (int?)stato);
                writer.WriteAttribute("Conteggio", Conteggio);
                writer.WriteAttribute("Montante", Montante);
            }

            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue) { this.SetFieldInternal(fieldName, fieldValue); }
            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer) { this.XMLSerialize(writer); }

            ~InfoStato()
            {
                DMDObject.DecreaseCounter(this);
            }
        }


        public class CompareByAnnullato
            : IComparer, IComparer<CRigaStatisticaPerStato>
        {
            public CompareByAnnullato()
            {
                DMDObject.IncreaseCounter(this);
            }

            ~CompareByAnnullato()
            {
                DMDObject.DecreaseCounter(this);
            }

            public int Compare(CRigaStatisticaPerStato x, CRigaStatisticaPerStato y)
            {
                decimal v1 = x.get_Item(InfoStatoEnum.ANNULLATO).Montante;
                decimal v2 = y.get_Item(InfoStatoEnum.ANNULLATO).Montante;
                if (v1 < v2)
                    return 1;
                if (v1 > v2)
                    return -1;
                return 0;
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CRigaStatisticaPerStato)x, (CRigaStatisticaPerStato)y);
            }
        }

        public class CompareByContatto : IComparer, IComparer<CRigaStatisticaPerStato>
        {
            public CompareByContatto()
            {
                DMDObject.IncreaseCounter(this);
            }

            ~CompareByContatto()
            {
                DMDObject.DecreaseCounter(this);
            }

            public int Compare(CRigaStatisticaPerStato x, CRigaStatisticaPerStato y)
            {
                decimal v1 = x.get_Item(InfoStatoEnum.CONTATTO).Montante;
                decimal v2 = y.get_Item(InfoStatoEnum.CONTATTO).Montante;
                if (v1 < v2)
                    return 1;
                if (v1 > v2)
                    return -1;
                return 0;
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CRigaStatisticaPerStato)x, (CRigaStatisticaPerStato)y);
            }
        }

        public class CompareByLiquidato : IComparer, IComparer<CRigaStatisticaPerStato>
        {
            public CompareByLiquidato()
            {
                DMDObject.IncreaseCounter(this);
            }

            ~CompareByLiquidato()
            {
                DMDObject.DecreaseCounter(this);
            }

            public int Compare(CRigaStatisticaPerStato x, CRigaStatisticaPerStato y)
            {
                decimal v1 = x.get_Item(InfoStatoEnum.LIQUIDATO).Montante;
                decimal v2 = y.get_Item(InfoStatoEnum.LIQUIDATO).Montante;
                if (v1 < v2)
                    return 1;
                if (v1 > v2)
                    return -1;
                return 0;
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CRigaStatisticaPerStato)x, (CRigaStatisticaPerStato)y);
            }
        }

        public class CompareByAltriStati : IComparer, IComparer<CRigaStatisticaPerStato>
        {
            public CompareByAltriStati()
            {
                DMDObject.IncreaseCounter(this);
            }

            ~CompareByAltriStati()
            {
                DMDObject.DecreaseCounter(this);
            }

            public int Compare(CRigaStatisticaPerStato x, CRigaStatisticaPerStato y)
            {
                decimal v1 = x.get_Item(InfoStatoEnum.ALTRO).Montante;
                decimal v2 = y.get_Item(InfoStatoEnum.ALTRO).Montante;
                if (v1 < v2)
                    return 1;
                if (v1 > v2)
                    return -1;
                return 0;
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CRigaStatisticaPerStato)x, (CRigaStatisticaPerStato)y);
            }
        }

        public class CompareByKey : IComparer, IComparer<CRigaStatisticaPerStato>
        {
            public CompareByKey()
            {
                DMDObject.IncreaseCounter(this);
            }

            ~CompareByKey()
            {
                DMDObject.DecreaseCounter(this);
            }

            public int Compare(CRigaStatisticaPerStato x, CRigaStatisticaPerStato y)
            {
                return DMD.Strings.Compare(x.Descrizione, y.Descrizione, true);
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CRigaStatisticaPerStato)x, (CRigaStatisticaPerStato)y);
            }
        }

        [Serializable]
        public class CRigaStatisticaPerStato 
            : IComparable, IDMDXMLSerializable
        {
            public string Descrizione = "";
            public InfoStato[] m_Items;
            public string Tag;
            public int Tag1;

            public CRigaStatisticaPerStato()
            {
                DMDObject.IncreaseCounter(this);
                m_Items = new InfoStato[4];
                m_Items[0] = new InfoStato("Contatto", InfoStatoEnum.CONTATTO, 0, 0m);
                m_Items[1] = new InfoStato("Liquidato", InfoStatoEnum.LIQUIDATO, 0, 0m);
                m_Items[2] = new InfoStato("Annullato", InfoStatoEnum.ANNULLATO, 0, 0m);
                m_Items[3] = new InfoStato("Altri Stati", InfoStatoEnum.ALTRO, 0, 0m);
            }

            public InfoStato get_Item(InfoStatoEnum stato)
            {
                switch (stato)
                {
                    case InfoStatoEnum.ALTRO:
                        {
                            return m_Items[3];
                        }

                    case InfoStatoEnum.ANNULLATO:
                        {
                            return m_Items[2];
                        }

                    case InfoStatoEnum.LIQUIDATO:
                        {
                            return m_Items[1];
                        }

                    default:
                        {
                            return m_Items[0];
                        }
                }
            }

            public int CompareTo(CRigaStatisticaPerStato b)
            {
                return DMD.Strings.Compare(Descrizione, b.Descrizione, true);
            }

            int IComparable.CompareTo(object obj)
            {
                return CompareTo((CRigaStatisticaPerStato)obj);
            }

            void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue) { this.SetFieldInternal(fieldName, fieldValue); }
            void IDMDXMLSerializable.XMLSerialize(XMLWriter writer) { this.XMLSerialize(writer); }
            protected virtual void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Descrizione":
                        {
                            Descrizione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Tag":
                        {
                            Tag = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Tag1":
                        {
                            Tag1 = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "Items":
                        {
                            m_Items = (InfoStato[])DMD.XML.Utils.Serializer.ToArray<InfoStato>(fieldValue);
                            break;
                        }
                }
            }

            protected virtual void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Descrizione", Descrizione);
                writer.WriteAttribute("Tag", Tag);
                writer.WriteAttribute("Tag1", Tag1);
                writer.WriteTag("Items", m_Items);
            }

            ~CRigaStatisticaPerStato()
            {
                DMDObject.DecreaseCounter(this);
            }
        }

        [Serializable]
        public class CRigaStatisticaPerStatoCollection : CKeyCollection<CRigaStatisticaPerStato>
        {
            public CRigaStatisticaPerStatoCollection()
            {
            }

            public new CRigaStatisticaPerStato Add(string descrizione)
            {
                var item = new CRigaStatisticaPerStato();
                item.Descrizione = descrizione;
                Add("" + descrizione, item);
                // MyBase.Sort()
                return item;
            }
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */


    }
}